kratos/utilities/volume_inside_voxel_utility.cpp

```diff
@@ -0,0 +1,382 @@
+                
+                if(result == 1) {
+                    double Dist = Distance(ends[0], intersection);
+                    Distances.push_back(Dist);
+                }  
+            } 
+            Distances.push_back(Distance(ends[0],ends[1]));
+            std::sort(Distances.begin(),Distances.end());       //WOULD A SET BE MORE EFFICIENT?     
+            double edgePortion = VolumeInsideVoxelUtility::EdgeFilledPortion(Distances, ends);
+            volume += edgePortion/Edges.size();  
+            Distances.clear();              
+        }
+        return volume;
+    }
+
+    /***********************************************************************************
+     **********************************************************************************/
+
+    double VolumeInsideVoxelUtility::VoxelVolume2D(
+        const GeometryType& rFace,  
+        const GeometryArrayType& rTriangles     
+    ) {
+        double Area = 0;
+        GeometryArrayType Edges = rFace.GenerateEdges();
+        PointsArrayType Nodes = rFace.Points(); 
+        std::vector<std::pair<double,double>> MinDistanceToNode(Edges.size(),{0.5,0.5}); 
+        //each pair represents an edge and contains as first() the minimum distance between
+        //ends[1] and an intersection and as second() the minimum distance between ends[1] and an intersection 
+
+        std::vector<double> Length(Edges.size()); 
+        for(int i = 0; i < Edges.size(); i++) {
+            PointsArrayType ends = Edges[i].Points();
+            double l = Distance(ends[0], ends[1]);
+            Length[i] = l;
+        }
+
+        for (int i = 0; i < rTriangles.size(); i++) {
+            int Result = 0; 
+            array_1d<double,3> Intersection;
+            int j = 0;
+
+            /*We will iterate through the edges using a while loop, so that if a triangles intersects 2 edges (unlikely 
+            but possible), only one will be taken into account to create the matrixes. */
+            while(!Result && j < Edges.size()) { 
+                PointsArrayType ends = Edges[j].Points();
+                Result = IntersectionUtilities::ComputeTriangleLineIntersection(rTriangles[i],ends[0],ends[1],Intersection);
+
+                if (Result) {
+                    double Dist = Distance(ends[0], Intersection);
+                    if ( Dist < (MinDistanceToNode[j].first*Length[j])) {
+                        MinDistanceToNode[j].first = Dist/Length[j];
+                    } 
+
+                    double Dist2 = Distance(ends[1], Intersection);
+                    if (Dist2 < (MinDistanceToNode[j].second*Length[j])) {
+                        MinDistanceToNode[j].second = Dist2/Length[j];
+                    } 
+                }
+                j++;
+            }
+        }
+
+        std::vector<std::vector<double>> Neighbours{{3,1},{0,2},{1,3},{0,2}};  
```

Comment:
I'd declare it as a constexpr in the header. Otherwise you're allocating it at each iteration of the loop.

Reply:
Besides, I think you can use an `std::vector<array_1d<std::size_t, 2>>`

Reply:
Maybe I am wrong, but I think there is no way to declare a std::vector as a constexpr (due to the way the constructor allocates memory?)